File: simplex-core/src/main/java/org/storck/simplex/model/BlockchainNotarized.java
package org.storck.simplex.model;

import java.util.ArrayList;
import java.util.List;

/**
 * The NotarizedBlockchain class represents a blockchain that consists of
 * notarized blocks.
 *
 * @param <T> the type of the transactions stored in the blocks
 * @param blocks the list of notarized blocks in the blockchain
 */
public record BlockchainNotarized<T>(List<BlockNotarized<T>> blocks) {

    /**
     * Create the notarized blockchain.
     *
     * @param blocks the blocks that constitute the chain
     */
    public BlockchainNotarized {
        blocks = new ArrayList<>(blocks);
    }

    /**
     * Returns an immutable copy of the list of notarized blocks in the blockchain.
     *
     * @return an immutable copy of the list of notarized blocks in the blockchain
     */
    public List<BlockNotarized<T>> blocks() {
        return List.copyOf(blocks);
    }

    /**
     * Adds a notarized block to the blockchain.
     *
     * @param newBlock the notarized block to be added
     */
    public void addBlock(final BlockNotarized<T> newBlock) {
        blocks.add(newBlock);
    }
}

File: simplex-messages/src/main/scala/org/storck/simplex/message/BlockchainNotarized.scala
package org.storck.simplex.message

/**
 * The NotarizedBlockchain class represents a blockchain that consists of
 * notarized blocks.
 *
 * @param blocks the list of notarized blocks in the blockchain
 */
case class BlockchainNotarized(blocks: Seq[BlockNotarized]) {

  /**
   * Adds a notarized block to the blockchain.
   *
   * @param newBlock the notarized block to be added
   * @return a new BlockchainNotarized instance with the added block
   */
  def addBlock(newBlock: BlockNotarized): BlockchainNotarized = {
    BlockchainNotarized(blocks :+ newBlock)
  }

  /**
   * Returns the number of blocks in the blockchain as a convenience.
   *
   * @return the number of blocks in the blockchain
   */
  def height: Int = blocks.size
}

object BlockchainNotarized {
  /**
   * Create the notarized blockchain.
   *
   * @param blocks the blocks that constitute the chain
   */
  def apply(blocks: Iterable[BlockNotarized]): BlockchainNotarized = {
    new BlockchainNotarized(blocks.toSeq)
  }
}